### PR TITLE
FPGA: Improve the `fpga_compile` tutorial

### PR DESCRIPTION
**Pull request by KevinUTAT**
# Existing Sample Changes
## Description

The `fpga_compile` tutorial is updated to have 5 different parts: C++, functor-USM, Lambda-Buffer, functor-buffer and lambda-USM. It's also updated to have co-simulation support.

The update to the tutorial is to show how a simple program like vector addition can go from traditional C++ implementation to SYCL for FPGA in different coding styles (lambda vs functor) and interfaces (buffer-accessor vs USM).

## External Dependencies

No change.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### Linux regtests result
https://spetc.intel.com/testanalysis?trview=0&testRunIds=7451962&tapgsize=20 \
Tested with 2023.1 compiler targeting pac_s10_usm board
### Windows regtests result
https://spetc.intel.com/testanalysis?trview=0&testRunIds=7471517&tapgsize=20 \
Tested with 2023.1 targeting IP Authoring because ref boards on windows farm do not support USM

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md [hunk 1]
```diff
@@ -252,61 +289,65 @@ Notice that whether you target the FPGA emulator, FPGA simulator or FPGA hardwar
...
 
-Code samples are licensed under the MIT license. See [License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.
+Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt).
 
-Third-party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt).
+[1]: <https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/analyze-your-design.html>
+[2]: <https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-compilation-flags.html>
+[3]: <https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html>
+[4]: <https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html>
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md:353]: Insert the links inline rather than at the end

Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md:353]: i thought we liked to have a list of links at the end of code samples?

Comment (yuguen-intel) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md:353]: I'm not aware of any samples doing so - maybe I missed something

Comment (KevinUTAT) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md]: Change link style back to inline

**Review by yuguen-intel — APPROVED** [review]

**Review by tomlenth — APPROVED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part5-dpcpp_functor_buffers/src/vector_add.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+using namespace sycl;
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class VectorAddID;
+
+template <class AccA, class AccB, class AccC>
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part5-dpcpp_functor_buffers/src/vector_add.cpp:13]: Declaring the struct with template type arguments (then using declttype) works, but does not do the same thing the other examples do.   The challenge is the original arrays needs to be read-write.   But you are capturing a read-only reference in the a struct.    The code "works", but if the intention is to show that equivalence, than it does not accomplish that.

Comment (KevinUTAT) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part5-dpcpp_functor_buffers/src/vector_add.cpp]: OK, removing part5 to keep consistency across

**Review by whitepau — APPROVED** [review]

**Comment by yuguen-intel** [comment]
@pmpeter1 are you happy with this revision?

File: DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part2-dpcpp_functor_usm/src/vector_add.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        // create the device queue
+        sycl::queue q(selector);
+
+        auto device = q.get_device();
+
+        std::cout << "Running on device: "
+                << device.get_info<sycl::info::device::name>().c_str()
+                << std::endl;
+
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part2-dpcpp_functor_usm/src/vector_add.cpp:52]:
```suggestion

if (!device.has(aspect::usm_host_allocations)) {
     std::terminate();
}


```

Check for USM support and exit gracefully

**Review by jimmytwei — APPROVED** [review]

**Review by pmpeter1 — APPROVED** [review]

**Pull request merged**